assets/src/block-editor/QueryBlockExtension/QueryBlockExtension.js

```diff
@@ -0,0 +1,151 @@
+import {ToolbarButton, RadioControl, PanelBody} from '@wordpress/components';
+import {InspectorControls, BlockControls} from '@wordpress/block-editor';
+import {useCallback, useEffect, useMemo, useState} from '@wordpress/element';
+
+const {addFilter} = wp.hooks;
+const {__} = wp.i18n;
+
+const targetBlocks = [ 'core/query' ];
+
+export const setupQueryBlockExtension = function() {
+  const {createHigherOrderComponent} = wp.compose;
+
+  addFilter(
+    'editor.BlockEdit',
+    'planet4-blocks/overrides/query-controls',
+    createHigherOrderComponent(
+      BlockEdit => props => {
+        if (!targetBlocks.includes(props.name)) {
+          return (
+            <BlockEdit {...props} />
+          );
+        }
+
+        const {attributes, setAttributes} = props;
+
+        const isSelectedBlock = namespace => {
+          const selectedBlock = wp.data.select('core/block-editor').getSelectedBlock();
+          return (selectedBlock && selectedBlock.attributes.namespace === namespace);
+        }
+
+        const onChangeHandler = useCallback(view => {
+          setAttributes({
+            displayLayout: {
+              type: view,
+              columns: view === 'carousel' ? 6 : 3,
+            },
+            query: {
+              ...attributes.query,
+              perPage: view === 'carousel' ? 6 : 3,
+            },
+          });
+        }, []);
+
+        useEffect(() => {
+          const pattern = 'is-custom-layout-';
+          const {className, displayLayout} = attributes;
+
+          if (typeof className !== 'undefined') {
+            setAttributes({
+              className: (className.includes(pattern)) ?
+                className.replace(/\is-custom-layout-.*/, `${pattern}${displayLayout.type}`) :
+                `${className} ${pattern}${displayLayout.type}`,
+            });
+          }
+        }, [
+          attributes,
+          setAttributes,
+        ]);
+
+        useEffect(() => {
+          // Reset in case of retrieving wrong values
+          if(attributes.displayLayout.type === 'flex' && attributes.displayLayout.columns === 6) {
+            setAttributes({
+              displayLayout: {
+                ...attributes.displayLayout,
+                columns: 3,
+              },
+              query: {
+                ...attributes.query,
+                perPage: 3,
+              },
+            });
+          }
+        }, [attributes.displayLayout.type])
+
+        useEffect(() => {
+          if (props.isSelected && isSelectedBlock('planet4-blocks/actions-list')) {
```

Comment:
Reduced the complexity of this.
Following UAT, the carousel is only applied to Actions List, and this block won't provide the `list` layout. Only `grid` and `carousel`.